cli/src/package.rs

```diff
@@ -1,36 +1,136 @@
-use std::{env::current_dir, fs, path::PathBuf};
-
+use crate::error::CliError;
 use directories::ProjectDirs;
+use futures::future::join_all;
+use modmark_core::Resolve;
+use serde_json;
+use std::{
+    env::current_dir,
+    fs::{self, create_dir_all, File},
+    io::copy,
+    path::PathBuf,
+};
 
-pub trait Resolve {
-    fn resolve(&self, path: &str) -> Option<Vec<u8>>;
-}
+pub struct PackageManager {}
 
-pub struct PackageManager;
 impl Resolve for PackageManager {
-    fn resolve(&self, path: &str) -> Option<Vec<u8>> {
-        let mut local_path = match current_dir() {
-            Ok(p) => p,
-            Err(_) => return None,
-        };
+    type Error = CliError;
+    fn resolve(&self, path: &str) -> Result<Vec<u8>, Self::Error> {
+        let out = tokio::runtime::Builder::new_multi_thread()
+            .enable_all()
+            .build()
+            .unwrap()
+            .block_on(resolve_package(path))?;
+
+        return Ok(out);
+    }
+    fn resolve_all(&self, paths: Vec<&str>) -> Vec<Result<Vec<u8>, Self::Error>> {
+        let out = tokio::runtime::Builder::new_multi_thread()
+            .enable_all()
+            .build()
+            .unwrap()
+            .block_on(resolve_packages(paths));
+
+        return out;
+    }
+}
+
+async fn resolve_packages(paths: Vec<&str>) -> Vec<Result<Vec<u8>, CliError>> {
+    let futures = paths.iter().map(|path| resolve_package(path));
 
-        local_path.push(PathBuf::from(path));
+    return join_all(futures).await;
+}
+
+async fn resolve_package(path: &str) -> Result<Vec<u8>, CliError> {
+    let splitter = path.split_once(":");
 
-        let mut cache_path = match ProjectDirs::from("org", "modmark", "packages") {
-            Some(path) => path.cache_dir().to_path_buf(),
-            None => return None,
+    let Some((specifier, package_path)) = splitter else {
+           return fetch_local(path);
         };
 
-        cache_path.push(PathBuf::from(path));
+    match specifier {
+        "https" => return fetch_url("https://www.", package_path).await,
+        "pkgs" => return fetch_registry(package_path).await,
+        other => return Err(CliError::Specifier(other.to_string())),
+    }
+}
+
+async fn fetch_url(url_root: &str, package_path: &str) -> Result<Vec<u8>, CliError> {
+    let mut cache_path = match ProjectDirs::from("org", "modmark", "packages") {
+        Some(path) => path.cache_dir().to_path_buf(),
+        None => return Err(CliError::Cache),
+    };
+
+    cache_path.push(PathBuf::from(&package_path));
+
+    let mut path = cache_path.clone();
+    path.pop();
+
+    create_dir_all(PathBuf::from(&path))?;
```

Comment:
This `let mut path ... path.pop ... pathbuf::from ...` thing to get the parent can be replaced with just `path.parent()` when using it, and here is the only spot you are using it,